miden-crypto/src/merkle/smt/large_forest/history.rs

```diff
@@ -0,0 +1,710 @@
+    /// avoid costly reallocations in the fast path.
+    ///
+    /// We use a [`VecDeque`] instead of a [`Vec`] or [`alloc::collections::LinkedList`] as we
+    /// estimate that the vast majority of removals will be the oldest entries as new ones are
+    /// pushed. This means that we can optimize for those removals along with indexing performance,
+    /// rather than optimizing for more rare removals from the middle of the sequence.
+    deltas: VecDeque<Delta>,
+}
+
+impl History {
+    /// Constructs a new history container, containing at most `max_count` historical versions for
+    /// a tree.
+    #[allow(dead_code)] // Temporary: this is internal code being merged incrementally
+    #[must_use]
+    pub fn empty(max_count: usize) -> Self {
+        let deltas = VecDeque::with_capacity(max_count);
+        Self { max_count, deltas }
+    }
+
+    /// Gets the maximum number of versions that this history can store.
+    #[allow(dead_code)] // Temporary: this is internal code being merged incrementally
+    #[must_use]
+    pub fn max_versions(&self) -> usize {
+        self.max_count
+    }
+
+    /// Gets the current number of versions in the history.
+    #[allow(dead_code)] // Temporary: this is internal code being merged incrementally
+    #[must_use]
+    pub fn num_versions(&self) -> usize {
+        self.deltas.len()
+    }
+
+    /// Returns `true` if a new entry can be added without removing the oldest, and `false`
+    /// otherwise.
+    #[allow(dead_code)] // Temporary: this is internal code being merged incrementally
+    #[must_use]
+    fn can_add_version_without_removal(&self) -> bool {
+        self.num_versions() < self.max_versions()
+    }
+
+    /// Adds a version to the history with the provided `root` and represented by the changes from
+    /// the current tree given in `nodes` and `leaves`.
+    ///
+    /// If adding this version would result in exceeding `self.max_count` historical versions, then
+    /// the oldest of the versions is automatically removed and returned. If no version is pruned,
+    /// the method returns [`None`].
+    ///
+    /// # Gotchas
+    ///
+    /// When constructing the `nodes` and `leaves`, keep in mind that those collections must contain
+    /// entries for the **default value of a node or leaf** at any position where the tree was
+    /// sparse in the state represented by `root`. If this is not done, incorrect values may be
+    /// returned.
+    ///
+    /// This is necessary because the changes are the _reverse_ from what one might expect. Namely,
+    /// the changes in a given version `v` must "_revert_" the changes made in the transition from
+    /// version `v` to version `v + 1`.
+    #[allow(dead_code)] // Temporary: this is internal code being merged incrementally
+    pub fn add_version(
+        &mut self,
+        root: Word,
+        nodes: NodeChanges,
+        leaves: LeafChanges,
+    ) -> Option<Delta> {
```

Comment:
A couple of potential changes to this API:

- We could take a `MutationSet` instead of `root`, `nodes`, and `leaves` as for our specific use case, updates will be generated via mutation sets.
- I would also pass in a `version` so that we could associate a given mutation set with a version. This should make pruning of history a bit easier in the future.

Combining the above, I think the signature could look something like this:
```rust
pub fn add_version(
    &mut self,
    version: u32,
    updates: MutationSet<SMT_DEPTH, Word, Word>,
) -> Option<Delta> {
    ...
}
```